### PR TITLE
Reimplement `MPS` and `MPO`

### PR DESCRIPTION
**Pull request by mofeing**
To do on subsequent PRs:

- #234
- #235
- #236
- #238
- #239
- #240
- #243
- #244

### CODE REVIEW

File: src/MPS.jl [hunk 1]
```diff
@@ -0,0 +1,330 @@
...
+Ansatz(tn::MPS) = tn.tn
+
+Base.copy(x::MPS) = MPS(copy(Ansatz(x)), form(x))
+Base.similar(x::MPS) = MPS(similar(Ansatz(x)), form(x))
+Base.zero(x::MPS) = MPS(zero(Ansatz(x)), form(x))
+
+defaultorder(::Type{MPS}) = (:o, :l, :r)
+boundary(::MPS) = Open()
+form(tn::MPS) = tn.form
+
```
Comment (starsfordummies) [inline src/MPS.jl:22]: """ Builds an MPS from a list of input tensors. Default ordering for the legs is (physical, virtual left, virtual right) - can be specified with the `order` keyword argument. 
Does/Does not? check for consistent dimension of input tensor legs
Does/Does not? bring to any canonical form 
"""

File: src/MPS.jl [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+    delete_Λ && delete!(TensorNetwork(tn), Λᵢ)
+
+    return tn
+end
+@kwmethod contract(tn::MPS; between) = contract(tn; between, direction=:left, delete_Λ=true)
+@kwmethod contract!(tn::MPS; between) = contract!(tn; between, direction=:left, delete_Λ=true)
+@kwmethod contract(tn::MPS; between, direction) = contract(tn; between, direction, delete_Λ=true)
+@kwmethod contract!(tn::MPS; between, direction) = contract!(tn; between, direction, delete_Λ=true)
+
+# TODO rename it to `lanes`?
```
Comment (starsfordummies) [inline src/MPS.jl:178]: please don't

File: src/MPS.jl [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+    elseif dir === :right && at == Site(nlanes(ψ); dual=isdual(at))
+        return nothing
+    elseif dir ∈ (:left, :right)
+        return inds(ψ; bond=(at, sites(ψ, at; dir)))
+    else
+        throw(ArgumentError("Unknown direction for MPS = :$dir"))
+    end
+end
+
+function isisometry(ψ::MPS, site; dir, atol::Real=1e-12)
```
Comment (starsfordummies) [inline src/MPS.jl:206]: """ Checks whether the tensor at `site` of the MPS `psi` is an isometry in the direction `dir = (left|right?)?` (up to tolerance `atol`). 
Eg. for an MPS in right canonical form, isisometry(psi, n; dir=left) should return `true`
"""

File: src/MPO.jl [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+Ansatz(tn::MPO) = tn.tn
+
+Base.copy(x::MPO) = MPO(copy(Ansatz(x)), form(x))
+Base.similar(x::MPO) = MPO(similar(Ansatz(x)), form(x))
+Base.zero(x::MPO) = MPO(zero(Ansatz(x)), form(x))
+
+defaultorder(::Type{MPO}) = (:o, :i, :l, :r)
+boundary(::MPO) = Open()
+form(tn::MPO) = tn.form
+
```
Comment (starsfordummies) [inline src/MPO.jl:26]: """ Builds an MPO from a list of input tensors. Default ordering for the legs is (physical out, physical in, virtual left, virtual right) - can be specified with the order keyword argument.
Does/Does not? check for consistent dimension of input tensor legs
"""

File: src/MPS.jl [hunk 5]
```diff
@@ -0,0 +1,335 @@
...
+    arrays[n] = reshape(arrays[n], p, p)
+
+    return MPS(arrays; order=(:l, :o, :r))
+end
+
+# TODO deprecate contract(; between) and generalize it to AbstractAnsatz
+"""
+    Tenet.contract!(tn::MPS; between=(site1, site2), direction::Symbol = :left, delete_Λ = true)
+
+For a given [`MPS`](@ref) tensor network, contracts the singular values Λ between two sites `site1` and `site2`.
```
Comment (starsfordummies) [inline src/MPS.jl:154]: contracts the tensors at site `site1` and `site2` with the SV between them ? que pasa si site1 y site2 are not adjacent?

File: src/MPO.jl [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+abstract type AbstractMPO <: AbstractAnsatz end
+
+"""
+    MPO <: AbstractAnsatz
+
+A Matrix Product Operator (MPO) [`Ansatz`](@ref) Tensor Network.
+"""
+struct MPO <: AbstractAnsatz
+    tn::Ansatz
+    form::Form
```
Comment (starsfordummies) [inline src/MPO.jl:14]: que es form?

File: src/MPO.jl [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+
+    # reshape boundary sites
+    arrays[1] = reshape(arrays[1], ip, op, min(χ, ip * op))
+    arrays[n] = reshape(arrays[n], min(χ, ip * op), ip, op)
+
+    # TODO order might not be the best for performance
+    return MPO(arrays; order=(:l, :i, :o, :r))
+end
+
+# TODO change it to `lanes`?
```
Comment (starsfordummies) [inline src/MPO.jl:119]: please don't

**Review by starsfordummies — APPROVED** [review]
some docs siusplau

**Comment by mofeing** [comment]
@starsfordummies gimme a break jajaja docs are on the way

File: src/MPO.jl (hunk 6 above)
Comment (mofeing) [reply src/MPO.jl:14]: it's the field containing the canonical form trait

imagine it as the field where you store the type of canonical form you're in and extra information about it (e.g. the orthogonality center in the case of the `MixedCanonical` form)

**Comment by starsfordummies** [comment]
just copy and paste my suggestions and fill the gaps, make this old man happy 🥺

File: src/MPO.jl (hunk 6 above)
Comment (starsfordummies) [reply src/MPO.jl:14]: ajá but why in MPO digo

Comment (mofeing) [reply src/MPO.jl:14]: por lo mismo, puedes usar los mismo canonical forms en MPO

File: src/MPS.jl (hunk 5 above)
Comment (mofeing) [reply src/MPS.jl:289]: one option is that if `site1` and `site2` are not neighbors (i.e. there is no edge connecting them in the lattice), then this should error
but for some stuff i'm interested in contracting non-contiguous tensors, so might only print a warning

File: src/Ansatz.jl [hunk 8]
```diff
@@ -192,13 +192,49 @@ Contract the virtual bond between two [`Site`](@ref)s in a [`AbstractAnsatz`](@r
...
+Like [`mixed_canonize!`](@ref), but returns a new Tensor Network instead of modifying the original one.
+"""
+mixed_canonize(tn::AbstractAnsatz, args...; kwargs...) = mixed_canonize!(deepcopy(tn), args...; kwargs...)
+
 canonize_site(tn::AbstractAnsatz, args...; kwargs...) = canonize_site!(deepcopy(tn), args...; kwargs...)
 
+"""
+    isisometry(tn::AbstractAnsatz, site; kwargs...)
+
+Check if the tensor at a given [`Site`](@ref) in a [`AbstractAnsatz`](@ref) Tensor Network is an isometry.
```
Comment (starsfordummies) [inline src/Ansatz.jl:230]: should specify the direction argument here ? otherwise it's a bit confusing

**Review by starsfordummies — APPROVED** [review]

File: src/Ansatz.jl (hunk 8 above)
Comment (mofeing) [reply src/Ansatz.jl:230]: mmm do you mean document `dir` kwarg? probably yeah, I'm still thinking around on how to encode directionality on `Lattice` for this so we can generalize to PEPS and Tree TNs

but meanwhile, I will put that `dir` kwarg in the docstring

for once, thanks @starsfordummies

File: test/integration/ChainRules_test.jl [hunk 9]
```diff
@@ -205,4 +205,24 @@
...
+
+    @testset "MPS" begin
+        tn = MPS([ones(2, 2), ones(2, 2, 2), ones(2, 2)])
+        # test_frule(MPS, Ansatz(tn), form(tn))
+        test_rrule(MPS, Ansatz(tn), form(tn))
+
+        # TODO reenable periodic MPS
+        # tn = MPS([ones(2, 2, 2), ones(2, 2, 2), ones(2, 2, 2)])
+        # test_frule(Chain, Quantum(tn), Periodic())
+        # test_rrule(Chain, Quantum(tn), Periodic())
```
[the remark this answers is not in the document]
Comment (mofeing) [reply test/integration/ChainRules_test.jl:217]: good point

**Review by jofrevalles — APPROVED** [review]

**Pull request merged**